bundle/terranova/deploy_mutator.go

```diff
@@ -0,0 +1,255 @@
+package terranova
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/bundle/terranova/tnresources"
+	"github.com/databricks/cli/bundle/terranova/tnstate"
+	"github.com/databricks/cli/libs/dagrun"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/cli/libs/structdiff"
+	"github.com/databricks/databricks-sdk-go"
+)
+
+// How many parallel operations (API calls) are allowed
+const defaultParallelism = 10
+
+type terranovaDeployMutator struct{}
+
+func TerranovaDeploy() bundle.Mutator {
+	return &terranovaDeployMutator{}
+}
+
+func (m *terranovaDeployMutator) Name() string {
+	return "TerranovaDeploy"
+}
+
+func (m *terranovaDeployMutator) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	client := b.WorkspaceClient()
+
+	g := dagrun.NewGraph[tnstate.ResourceNode]()
+
+	_, err := dyn.MapByPattern(
+		b.Config.Value(),
+		dyn.NewPattern(dyn.Key("resources"), dyn.AnyKey(), dyn.AnyKey()),
+		func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			section := p[1].Key()
+			name := p[2].Key()
+			g.AddNode(tnstate.ResourceNode{
+				Section: section,
+				Name:    name,
+			})
+
+			// TODO: Scan v for references and use g.AddDirectedEdge to add dependency
+			return v, nil
+		},
+	)
+
+	if g.Size() == 0 {
+		// Avoid creating state file if nothing to deploy
+		return nil
+	}
+
+	var diags diag.SafeDiagnostics
+
+	err = g.Run(defaultParallelism, func(node tnstate.ResourceNode) {
+		// TODO: if a given node fails, all downstream nodes should not be run. We should report those nodes.
+		// TODO: ensure that config for this node is fully resolved at this point.
+
+		config, ok := b.GetResourceConfig(node.Section, node.Name)
+		if !ok {
+			diags.AppendErrorf("internal error: cannot get config for %s", node)
+			return
+		}
+
+		d := Deployer{
+			client:       client,
+			db:           &b.ResourceDatabase,
+			section:      node.Section,
+			resourceName: node.Name,
+		}
+
+		err = d.Deploy(ctx, config)
+		if err != nil {
+			diags.AppendError(err)
+			return
+		}
+	})
+	if err != nil {
+		diags.AppendError(err)
+	}
+
+	err = b.ResourceDatabase.Finalize()
+	if err != nil {
+		diags.AppendError(err)
+	}
+
+	return diags.Diags
+}
+
+type Deployer struct {
+	client       *databricks.WorkspaceClient
+	db           *tnstate.TerranovaState
+	section      string
+	resourceName string
+}
+
+func (d *Deployer) Deploy(ctx context.Context, inputConfig any) error {
+	err := d.deploy(ctx, inputConfig)
```

Comment:
Can be inlined as it's not used anywhere else

Reply:
This is done on purpose, to add prefix with resource name in a single place.